doc/rst/technotes/foreach.rst

```diff
@@ -0,0 +1,38 @@
+.. _readme-foreach:
+
+====================
+The `foreach` Loop
+====================
+
+A ``foreach`` loop signifies that loop iterations are order-independent. This
+enables the compiler to vectorize the loop. In contrast, a ``for`` loop is
+order-dependent, and a ``forall`` loop can create parallel tasks and distribute
+iterations based on the iterator.
+
+The syntax of the ``foreach`` loop is similar to the other ``for``-like loops:
+
+.. code-block:: chapel
+
+    var A: [1..N] int;
+    foreach i in A.domain do
+      A[i] += 1;
+
+Vectorizability of a ``foreach`` loop will be determined transitively based on
+vectorizability of the iterator it runs over. All default range, domain and
+array iterators are vectorizable (i.e. implemented using ``foreach``).
+Therefore, the snippet above will be vectorized when run on a suitable
+architecture.
+
+Same transitivity applies to vectorizability of ``forall`` loops, as well.
+
+.. code-block:: chapel
+
+    var Dom = {1..N} dmapped Block({1..N});
+    var A: [Dom] int;
+    forall i in A.domain do
+      A[i] += 1;
+
+The ``forall`` above will distribute the execution to all the locales and all
+the tasks within them. As the default parallel iterators for ``A.domain`` is
+implemented using ``foreach``, tasks executing parts of this loop will benefit
+from vectorization.
```

Comment:
I think we need to add something at the end here indicating the current implementation status, something like:
 * implementation not yet making full use of this vectorization hinting
 * intend to add `with` clauses similar to `forall` to the `foreach` loop